### PR TITLE
SelfHostedEtcd: Add kenc as DS in master nodes

### PR DESCRIPTION
**Pull request by hongchaodeng**
`kenc` is an iptables checkpoint tool.
With it, service IP connectivity on master nodes can endure reboot.

### CODE REVIEW

**Comment by hongchaodeng** [comment]
Needed:
- [ ] enable disruption test suite to test this PR

File: pkg/asset/internal/templates.go [hunk 1]
```diff
@@ -198,6 +198,43 @@ spec:
...
+      labels:
+        k8s-app: kenc
+      annotations:
+        checkpointer.alpha.coreos.com/checkpoint: "true"
+    spec:
+      nodeSelector:
+        master: "true"
+      hostNetwork: true
+      containers:
+      - image: gcr.io/coreos-k8s-scale-testing/kenc
```
Comment (aaronlevy) [inline pkg/asset/internal/templates.go]: Prior to merging we should have a quay/coreos repo + release tag

File: pkg/asset/internal/templates.go [hunk 2]
```diff
@@ -198,6 +198,43 @@ spec:
...
+        name: kenc
+        securityContext:
+          privileged: true
+        volumeMounts:
+        - mountPath: /etc/kubernetes/selfhosted-etcd
```
Comment (aaronlevy) [inline pkg/asset/internal/templates.go]: For pod checkpointing we use:

`/etc/kubernetes/checkpoint-configmaps`
`/etc/kubernetes/checkpoint-secrets`

Maybe for consistency we use `/etc/kubernetes/checkpoint-iptables`?

Comment (hongchaodeng) [inline pkg/asset/internal/templates.go]: There is an option to configure it. Let me make it.

Comment (hongchaodeng) [inline pkg/asset/internal/templates.go]: Actually we only need to change hostPath..

**Comment by aaronlevy** [comment]
We may want to consider adding a flock around the execution of this pod - unless you're sure it's safe to potentially have two running at once. The checkpointing process does not guarantee that parent/child have some overlap. Example with api-server: https://github.com/kubernetes-incubator/bootkube/blob/master/pkg/asset/internal/templates.go#L151-L155

**Comment by hongchaodeng** [comment]
> The checkpointing process does not guarantee that parent/child have some overlap

What does it mean?

**Comment by aaronlevy** [comment]
Sorry, not very helpful of description:

Both the original pod and the checkpointed copy of that pod could end up running at the same time. The checkpointer does not enforce that only one is running at a time, and there likely will be overlap.

**Comment by hongchaodeng** [comment]
@aaronlevy 

Thanks for explaining. I think it's best practice for all such pods to do "flock", although for kenc overlapping doesn't matter. I will change it shortly.

**Comment by hongchaodeng** [comment]
Added flock

**Comment by hongchaodeng** [comment]
@aaronlevy 
Any other comment on this?

**Comment by aaronlevy** [comment]
lgtm. Is this ready to be merged?

**Comment by hongchaodeng** [comment]
@aaronlevy 

Yes. Ready for merge. We have also tested it.

Also see https://github.com/kubernetes-incubator/bootkube/pull/356#issuecomment-284857226 .
Enabling destruction test will be delayed because we need etcd operator to do a new release.

**Comment by aaronlevy** [comment]
Gonna wait for PR test before merge

**Pull request merged**